aten/src/ATen/core/stack.h

```diff
@@ -77,9 +77,21 @@ static inline void pop(Stack& stack, Types&... args) {
   (void)result;
   drop(stack, N);
 }
+template <typename Type>
+static inline void push_one(Stack& stack, Type&& arg) {
+  stack.emplace_back(std::forward<Type>(arg));
+}
+
+static inline void push_one(Stack& stack, c10::TensorOptions options) {
+  stack.emplace_back(c10::typeMetaToScalarType(options.dtype()));
+  stack.emplace_back(options.layout());
+  stack.emplace_back(options.device());
+  stack.emplace_back(options.pinned_memory());
+}
+
 template <typename... Types>
 static inline void push(Stack& stack, Types&&... args) {
-  (void)std::initializer_list<int>{(stack.emplace_back(std::forward<Types>(args)), 0)...};
+  (void)std::initializer_list<int>{(push_one(stack, args), 0)...};
```

Comment:
do you think we could add a fast path? something like
```
// enable_if all ivalue constructible
stack.reserve(stack.size() + sizeof...(Types));
auto v = { IValue(args)... };
stack.insert(stack.end(), v.begin(), v.end());
```

Reply:
reason: the realloc insert is taking up a non-negligible amount of time in profiling (2%)

Reply:
Yes, I'll do this in a follow up.

Reply:
@bwasti I looked into applying this optimization and I don't think we can safely do it (in fact, the `reserve` in `push_list_elements` is also a bit suspect).  The problem is that reserve can cause quadratic behavior when called in a loop (http://slashslash.info/2013/11/capacity-members-for-vector-reserve/). And in general, the point of putting arguments on a stack is so that we CAN reuse the stack for nested calls without requiring more memory allocations. So reserve is not the right thing here.